kuma/wiki/templatetags/ssr.py

```diff
@@ -0,0 +1,128 @@
+                                    stdin=subprocess.PIPE,
+                                    stdout=subprocess.PIPE,
+                                    stderr=sys.stderr)
+    return renderer
+
+
+def server_side_render(data):
+    """
+    Pre-render the React UI to HTML and output it in a <div>, and then
+    also pass the necessary serialized state in a <script> so that
+    React on the client side can sync itself with the pre-rendred HTML.
+
+    If any exceptions are thrown during the server-side rendering, we
+    fall back to client-side rendering instead.
+    """
+    start = time.time()
+
+    # Try server side rendering
+    try:
+        renderer = _get_renderer()
+
+        # To server-side render, we send the JSON-encoded data to
+        # the child process on its stdin stream, and then send
+        # a line of four hyphens to mark the end of the data.
+        renderer.stdin.write(json.dumps(data))
+        renderer.stdin.write('\n----\n')
+        renderer.stdin.flush()
+
+        # The renderer process responds by writing HTML to its stdout
+        # stream. But first, it writes a line of text that it uses
+        # as a boundary (in the MIME multipart sense). So we read the
+        # boundary line, then read lines from the child process until
+        # we see the boundary line again.
+        boundary = renderer.stdout.readline().decode('utf-8')
+        renderedHTML = ''
+        while True:
+            line = renderer.stdout.readline().decode('utf-8')
+            if line == boundary:
+                break
+            renderedHTML += line
+
+        # Even though we've got fully rendered HTML now, we still need
+        # to send the document data along with it so that React can sync
+        # its state on the client side with what is in the HTML.
+        # Fortunately, however, it turns out not to be necessary to
+        # duplicate the biggest parts (the HTML strings) of that data,
+        # so we can delete those from the data now
+        partial = data.copy()
+        partial['bodyHTML'] = ''
+        partial['tocHTML'] = ''
+        partial['quickLinksHTML'] = ''
+        # Convert to JSON and be sure the string "</script>" does not appear
+        serializedState = json.dumps(partial).replace('</', '<\\/')
+
+        # Is there a better way to write log messages from a template tag?
+        # On my dev server, rendering takes 30-50ms once the server is
+        # already running and 150-200ms when the server needs to start
+        print "Server side rendered in {}s".format(time.time() - start)
+
+        return OUTPUT_TEMPLATE.format(renderedHTML, serializedState)
+
+    except Exception as e:
+        # If server side rendering fails, fall back to client-side
+        # Is there a better way to log this?
+        print "SSR failed after {}s".format(time.time() - start)
+        print e
+        print "Falling back to client side rendering..."
+        return client_side_render(data)
```

Comment:
Actually, I think we should eliminate this. `except Exception` is generally a code-smell because it's too big and too broad. Again, I don't know how "serious" this code is on a scale of prototype vs. production ready. 

*if* any exception happens on the 49 lines of code (that it covers) I'd rather we let it happen and learn from it. It should end up in Sentry and since we have weeks/months until this needs to go into Prod proper, we can fail forward by patching unexpected errors deliberately. If anything, we could refactor it to look like this:


```python
renderer = _get_renderer()
renderer.stdin.write(json.dumps(data))
...

# The reading from stdout the renderer can sometimes trigger bla bla yada yada
try:
    while True:
       ...reading from stdout...
except SomeSpecificSubprocessError as exception:
    return client_side_data(data, exception=exception)

...continue with the rest...
```

Reply:
You make a good point. I'll wait and see what Ryan things and whether we decide to move forward with this approach before deciding whether to remove or keep  the except clause.

If I'm able to use childprocess32 to apply timeouts to the read and write calls, then I think it might be legit to catch those. And probably kill the child process if it happens.

Is there a way we can handle an exception *and* report it to Sentry?